src/radical/pilot/agent/scheduler/continuous.py

```diff
@@ -255,50 +262,67 @@ def _get_node_maps(self, cores, gpus, threads_per_proc):
+        # For threads, if the entry is not specified in the description, then 
+        # assume it is one thread per process.
         requested_procs  = cud['cpu_processes']
-        threads_per_proc = cud['cpu_threads']
+        threads_per_proc = cud.get('cpu_threads',1)
         requested_gpus   = cud['gpu_processes']
 
-        if not threads_per_proc:
-            threads_per_proc = 1
+        # The total number of requested logical cores is equal to the number of
+        # processes times the number of threads per process.
 
         requested_cores = requested_procs * threads_per_proc
 
-        # we want the allocation to fit on a single node - make sure this is
-        # actually possible
+        # The allocation has to fit in a single node. If this is true contnue and
+        # try to find a possible allocation, otherwise raise a value error exception.
         if  requested_cores > self._lrms_cores_per_node or \
             requested_gpus  > self._lrms_gpus_per_node     :
             raise ValueError('Non-mpi unit does not fit onto single node')
 
-        # ok, we can go ahead and try to find a matching node.
+        # Initiallize core and gpu list that the unit may allocate. Also initialize
+        # node name and node id variables.
         cores     = list()
         gpus      = list()
         node_name = None
         node_uid  = None
+
         for node in self.nodes:  # FIXME optimization: iteration start
 
-            # attempt to allocate the required number of cores and gpus on this
+            # attempt to find the required number of cores and gpus on this
             # node - do not allow partial matches.
             cores, gpus = self._alloc_node(node, requested_cores, requested_gpus,
                                            partial=False)
 
+            # if the nessecary number of cores and GPUs is found, break the loop.
+            # Else check the next node.
             if  len(cores) == requested_cores and \
                 len(gpus)  == requested_gpus      :
                 # we are done
                 node_uid  = node['uid']
                 node_name = node['name']
                 break
 
+        # In the case there is not enough space for this unit in any node, return
+        # None to signal that no allocation was possible.
         if not cores and not gpus:
-            return None  # signal failure
+            return None
 
-        # we have to communicate to the launcher where exactly processes are to
-        # be placed, and what cores are reserved for application threads.
+        # A suitable allocation was found and the core and gpu map, core and gpu
+        # ids, should be return to the launcher. This way the processes will be
+        # placed in cores with distance equal to the number of threads, leaving
+        # enough space between them.
         core_map, gpu_map = self._get_node_maps(cores, gpus, threads_per_proc)
```

Comment:
`distance` is actually not correct.  Assume we want `n_procs=6, threads_per_procs=2` and find cores `0, 1, 2, 3, 4, 5, 7, 8` - then they are not equal-distant.
Maybe: 
```
 +        # A suitable allocation was found and the core and gpu map, core and gpu
￼ +        # ids, should be return to the launcher, informing about the exact placement
 +        # of processes and threads, respectively. 
```
?

Reply:
Yes, threads can go to any cores, they don't need to be next to each other.  The only limitation really is that they live on the same OS instance - which we assume to be the case for nodes in Beowulf-like clusters.

Locating threads on the same physical CPU (in multi-CPU nodes) has some performance benefits, but we do not explicitly attempt to exploit that.